### PR TITLE
vim-patch:8.0.{1459,1460,1461,1463},8.1.{0602,0604,1291},8.2.{0189,0876,0909,1411}: chdir and DirChanged related patches

### LINKED ISSUE #15939: Behavior of `:cd -`, `:tcd -`, `:lcd -` is different from Vim
### Neovim version (nvim -v)

NVIM v0.6.0-dev+423-ge06936125a

### Vim (not Nvim) behaves the same?

no, vim 8.2.3441

### Operating system/version

Arch Linux

### Terminal name/version

kitty 0.23.1

### $TERM environment variable

xterm-kitty

### Installation

AUR neovim-git

### How to reproduce the issue

1. Create three directories `/tmp/a`, `/tmp/b`, `/tmp/c`
2. Run `nvim --clean`
3. `:tcd /tmp/a`
4. `:tcd /tmp/b`
5. `:lcd /tmp/c`
6. `:tcd -`

### Expected behavior

The last `:tcd -` goes to `/tmp/a`.

### Actual behavior

The last `:tcd -` goes to `/tmp/b`

Change `:tcd` and `:lcd` to any 2-permutation of `:cd`, `:tcd`, `:lcd`, and the bug also happens.

Related Vim patch: [8.2.0909](https://github.com/vim/vim/commit/002bc79991286934a9593b80635c27d4238cdfc4). Neovim implemented `:tcd -` and `:lcd -` much earlier than Vim, but only restores `prev_dir` globally:
<https://github.com/neovim/neovim/blob/c61a3865eea21f87ac17719ba226ad556b5a11a6/src/nvim/ex_docmd.c#L7699>

### PR DESCRIPTION
**Pull request by zeertzjq**
I did maintain the behavior of `v:event.changed_window`. I'm not sure exactly what should be maintained and what should not. I mainly changed these in this PR:
- Don't trigger `DirChanged` when `:lcd`, `:tcd`, `:cd` didn't change directory. (Relative path always triggers `DirChanged`, <https://github.com/vim/vim/issues/8978#issuecomment-940962535>.)
- Add `chdir()` function.
- Set `<amatch>` to `"auto"` instead of `"window"` if triggered by `'autochdir'`. `v:event.scope` is still `"window"`.
- Always call `slash_adjust` for `cwd` under MS Windows.
- Make the behavior of `:lcd -` and `:tcd -` match both Vim and Neovim's documentation.
- Change scope name `"tab"` to `"tabpage"` to match Vim's `"tabpage"` pattern. I [cannot find](https://sourcegraph.com/search?q=context:global+DirChanged.*v:event&patternType=regexp&case=yes) any existing plugin that relies on the `"tab"` scope name, so I think this change is safe.
- Prevent `DirChanged` from being triggered by `sesdir` of `'sessionoptions'`. I think it shouldn't be triggered here, as it isn't triggered when changing back to `dirnow` either.
- Document related difference between Neovim and Vim in `vim_diff.txt`.
- Move `E5000`, `E5001`, `E5002` tags to `vim_diff.txt`.

`<amatch>` is previously undocumented, so I think only plugins that intend to support both Vim and Neovim use it. Neovim-only plugin have `v:event.scope` to use instead.

I think the "drop" pattern is N/A.

Vim patch 8.1.1218 also includes some changes to `getcwd()` and `haslocaldir()`, which I'm not sure about.

Close #15939.

### CODE REVIEW

**Review by justinmk — APPROVED** [review]
Nice! This is valuable, thank you.

**Comment by justinmk** [comment]
> skip test_autocmd.vim: depends on patch 8.0.1459

 https://github.com/vim/vim/commit/b7407d3fc9496f9048fb65ab17b5ba3444965c0e

can you explain why skip test_autocmd.vim ? Even if the test is N/A, we try to keep the tests near-identical. If a specific line is N/A, comment it out. If a block of logic is N/A, skip it with `if has('nvim')`

**Comment by zeertzjq** [comment]
Currently Neovim and Vim's pattern matching for `DirChanged` are a bit different: Neovim matches against the scope string:
<https://github.com/neovim/neovim/blob/a36c6e5df959867ff4041405f4098b7833281517/src/nvim/file_search.c#L1623>
whereas Vim has two other strings: "auto" and "drop". I think "auto" and "drop" should not be considered scopes, so I cannot decide whether Neovim's pattern matching for `DirChanged` should be changed to match Vim's

**Comment by zeertzjq** [comment]
> > skip test_autocmd.vim: depends on patch 8.0.1459
> 
> [vim/vim@b7407d3](https://github.com/vim/vim/commit/b7407d3fc9496f9048fb65ab17b5ba3444965c0e)
> 
> can you explain why skip test_autocmd.vim ? Even if the test is N/A, we try to keep the tests near-identical. If a specific line is N/A, comment it out. If a block of logic is N/A, skip it with `if has('nvim')`

The main problem is the `"auto"` pattern in `DirChanged` introduced in Vim patch 8.0.1459, which Neovim doesn't support. I can guard tests for the `"auto"` pattern with `if has('nvim')`.

**Comment by zeertzjq** [comment]
The Windows autocmd tests are failing because Neovim uses the path retrieved by `uv_cwd()` to as `<afile>` for `DirChanged` (backslash), whereas Vim uses the string passed to `chdir()` as-is (slash). IDK what is the best sokution to this.

**Comment by zeertzjq** [comment]
> The Windows autocmd tests are failing because Neovim uses the path retrieved by `uv_cwd()` to as `<afile>` for `DirChanged` (which always uses backslash as path separators), whereas Vim uses the string passed to `chdir()` as-is (which may use slash or backslash). IDK what is the best solution to this.

Maybe current solution is to add `CheckNotMSWindows`.

**Comment by zeertzjq** [comment]
> The Windows autocmd tests are failing because Neovim uses the path retrieved by `uv_cwd()` as `<afile>` for `DirChanged` (which always uses backslashes as path separators on Windows), whereas Vim uses the string passed to `chdir()` as-is (which may use slashes or backslashes). IDK what is the best solution to this.

I'm currently trying `slash_adjust`.

**Comment by zeertzjq** [comment]
`E5000`, `E5001`, `E5002` tags seem to fit better beside `getcwd()`. I moved them there.

**Comment by zeertzjq** [comment]
I didn't change Neovim's behavior of `getcwd()` and `haslocaldir()` to match Vim's, as I think these are breaking changes that should be introduced in a separate PR or not introduced at all. I just documented the differences in `vim_diff.txt`.

**Comment by janlazo** [comment]
Patches v8.0.{1459,1460,1461,1463} aren't addressed here. They should be resolved before (or on the same commit) as the v8.1.x patches for `DirChanged`.

**Comment by zeertzjq** [comment]
I did prevent `DirChanged` from triggering because of `sesdir` of `'sessionoptions'`. I'm not sure about this.

**Comment by janlazo** [comment]
You skipped `DO_AUTOCHDIR` changes.

**Comment by zeertzjq** [comment]
> You skipped `DO_AUTOCHDIR` changes.

Oh, indeed. I think I just have to remove the call to `do_autochdir()` in `close_buffer()`.

**Review by janlazo — APPROVED** [review]
LGTM

It was hard to follow because of the commit order. Some tests and docs depend on later patches.

**Comment by zeertzjq** [comment]
Oh, I included some test changes introduced in patch 8.1.2278 in 8.1.0602, because I cannot fully port 8.1.2278. I can split that out.
